torch/csrc/distributed/c10d/StoreCollectives.cpp

```diff
@@ -0,0 +1,191 @@
+#include <torch/csrc/distributed/c10d/Store.hpp>
+#include <torch/csrc/distributed/c10d/StoreCollectives.hpp>
+#include <chrono>
+#include <exception>
+#include <vector>
+
+namespace c10d {
+
+StoreCollectives::StoreCollectives(
+    c10::intrusive_ptr<::c10d::Store> store,
+    int rank,
+    int world_size)
+    : store_(store), rank_(rank), world_size_(world_size) {}
+
+void StoreCollectives::barrier(
+    const std::string& prefix,
+    std::chrono::milliseconds timeout,
+    bool blocking) {
+  StoreTimeoutGuard g{*store_, timeout};
+
+  auto num_members_key = prefix + "/num_members";
+  auto last_members_key = prefix + "/last_members";
+
+  auto idx = store_->add(num_members_key, 1);
+  store_->set(prefix + "/" + std::to_string(rank_), "joined");
```

Comment:
since you are doing a lot of string concatenation in these files, may want to consider looking to see if fmtlib help makes the code simpler + faster. It's already a dependency in PyTorch.